lib/wit.js

```diff
@@ -57,76 +62,89 @@ function Wit(opts) {
-      if (json.type === 'merge') {
-        json.type = 'action';
-        json.action = 'merge';
+      // backwards-compatibility with API version 20160516
+      if (type === 'merge') {
+        type = 'action';
+        action = 'merge';
       }
 
-      if (json.type === 'error') {
+      if (type === 'error') {
         throw new Error('Oops, I don\'t know what to do.');
       }
 
-      if (json.type === 'stop') {
+      if (type === 'stop') {
         return prevContext;
       }
 
       const request = {
         sessionId,
         context: clone(prevContext),
         text: message,
-        entities: json.entities,
+        entities
       };
 
-      if (json.type === 'msg') {
+      if (type === 'msg') {
         throwIfActionMissing(actions, 'send');
         const response = {
-          text: json.msg,
-          quickreplies: json.quickreplies,
+          text: msg,
+          quickreplies,
         };
         return actions.send(request, response).then((ctx) => {
           if (ctx) {
             throw new Error('Cannot update context after \'send\' action');
           }
+          throwIfNotLastIndex(index, this.sessions[sessionId].index);
           return this.converse(sessionId, null, prevContext).then(
-            continueRunActions(sessionId, message, prevContext, i - 1)
+            continueRunActions(sessionId, index, message, prevContext, i - 1)
           );
         });
-      } else if (json.type === 'action') {
-        const action = json.action;
+      } else if (type === 'action') {
         throwIfActionMissing(actions, action);
         return actions[action](request).then((ctx) => {
+          throwIfNotLastIndex(index, this.sessions[sessionId].index);
           const nextContext = ctx || {};
           return this.converse(sessionId, null, nextContext).then(
-            continueRunActions(sessionId, message, nextContext, i - 1)
+            continueRunActions(sessionId, index, message, nextContext, i - 1)
           );
         });
       } else {
-        logger.debug('unknown response type', json);
-        throw new Error('unknown response type ' + json.type);
+        logger.debug('unknown response type ' + type);
+        throw new Error('unknown response type ' + type);
       }
     }
   };
 
   this.runActions = (sessionId, message, context, maxSteps) => {
     if (!actions) throwMustHaveActions();
+
+    let {index = 0, text = ''} = this.sessions[sessionId] || {};
+    ++index;
+    text = text === '' ? message : text + ' ' + message;
+    this.sessions[sessionId] = { index, text };
+
     const steps = maxSteps ? maxSteps : DEFAULT_MAX_STEPS;
-    return this.converse(sessionId, message, context).then(
-      continueRunActions(sessionId, message, context, steps)
-    );
+    return this.converse(sessionId, text, context, index > 1).then(
+      continueRunActions(sessionId, index, text, context, steps)
+    ).then((ctx) => {
+      if (index === this.sessions[sessionId].index) {
```

Comment:
add a comment explaining cleanup.. maybe even declare a `cleanup` closure above that you invoke here